### PR TITLE
ValidatingAdmissionPolicy: support variables in TypeChecking

### LINKED ISSUE #122935: validatingadmissionpolicystatus reports error when using spec.variables
### What happened?

When I load a policy definition like this

```yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingAdmissionPolicy
metadata:
  name: loadbalancer.servicenow.net
spec:
  matchConstraints:
    resourceRules:
    - apiVersions: ["v1"]
      apiGroups:   [""]
      resources:   ["services"]
      operations:  ["CREATE", "UPDATE"]
  matchConditions:
  - name: "only-loadbalancer-services"
    expression: object.spec.type == "LoadBalancer"
  paramKind:
    apiVersion: control.servicenow.net/v1beta1
    kind: TenantPermission
  variables:
  - name: loadbalancerid
    expression: "'ccm.snowsk8s.servicenow.net/lb-service-id' in object.metadata.annotations ? object.metadata.annotations['ccm.snowsk8s.servicenow.net/lb-service-id'] : 'none'"
  validations:
    - expression: variables.loadbalancerid in params.permissions.loadbalancerid
      reason: Forbidden
      messageExpression: "'Namespace ' + string(namespaceObject.metadata.name) + ' does not have permissions to loadbalancer ' + string(variables.loadbalancerid)"
  failurePolicy: Fail
```

The object is updated with the following status
```yaml
status:
  observedGeneration: 2
  typeChecking:
    expressionWarnings:
    - fieldRef: spec.validations[0].expression
      warning: |
        /v1, Kind=Service: ERROR: <input>:1:1: undeclared reference to 'variables' (in container '')
         | variables.loadbalancerid in params.permissions.loadbalancerid
         | ^
    - fieldRef: spec.validations[0].messageExpression
      warning: |
        /v1, Kind=Service: ERROR: <input>:1:111: undeclared reference to 'variables' (in container '')
         | 'Namespace ' + string(namespaceObject.metadata.name) + ' does not have permissions to loadbalancer ' + string(variables.loadbalancerid)
         | ..............................................................................................................^
```

While this error is reported, the policy does seem to be fully functioning

```
// Try to claim someone elses loadbalancer in the a different namespace
k apply -f testing/someones-loadbalancer.yaml   
Error from server (Forbidden): error when creating "testing/someones-loadbalancer.yaml": services "idonthavealoadbalancer" is forbidden: ValidatingAdmissionPolicy 'loadbalancer.servicenow.net' with binding 'deny-non-owning-loadbalancer' denied request: Namespace workload1 does not have permissions to loadbalancer 123someoneelses123

// Try to claim a loadbalancer without any valid ID
 k apply -f testing/idonthavemyown.yaml
Error from server (Forbidden): error when creating "testing/idonthavemyown.yaml": services "soneone-elses-loadbalancer" is forbidden: ValidatingAdmissionPolicy 'loadbalancer.servicenow.net' with binding 'deny-non-owning-loadbalancer' denied request: Namespace workload1 does not have permissions to loadbalancer none

// A valid request
k apply -f testing/my-loadbalancer.yaml 
service/my-loadbalancer configured
```

This is similar to the documented examples
https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/#variable-composition

### What did you expect to happen?

No error reported on a valid spec

### How can we reproduce it (as minimally and precisely as possible)?

Create a a policy that uses variables
Observe the status of the newly created policy

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
k version
Client Version: v1.29.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.0
```

</details>

### Cloud provider

<details>
n/a
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>
kind
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by jiahuif**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR add support of Variable Composition to ValidatingAdmissionPolicy by:

- refactor the Type Checker to use the FilterCompiler mechanism instead of raw CEL compilation
- create an implementation of compiler that uses actual types instead of `Dyn`, for `object`, `oldObject`, and `params`
- wire in the `CompositionEnv`
- add test cases for using `variables`

#### Which issue(s) this PR fixes:

Fixes #122935

#### Does this PR introduce a user-facing change?

```release-note
ValidatingAdmissionPolicy now supports type checking policies that make use of `variables`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by jiahuif** [comment]
/retest
for flaky test.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/typechecking_test.go [hunk 1]
```diff
@@ -413,6 +413,95 @@ func TestTypeCheck(t *testing.T) {
...
+				},
+			},
+			assertions: []assertionFunc{
+				toHaveLengthOf(1),
+				toHaveFieldRef("spec.validations[0].expression"),
+				toContain("undefined field 'nonExisting'"),
+			},
+		},
+		{
+			name: "variables field wrong type",
```
Comment (cici37) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/typechecking_test.go:475]: I think it's worth to add another test case with type checking err from variable expression :)

Comment (jiahuif) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/typechecking_test.go:475]: Yep! However do note that the errors inside variables are delayed to runtime, as defined in the Variable Composition section of our KEP.

**Comment by jiahuif** [comment]
/retest
unrelated flake

**Comment by cici37** [comment]
@jiahuif This PR should be back ported to earlier version(At least 1.28 and 1.29). Thanks

**Comment by jiahuif** [comment]
There is a slight problem. The backport to 1.28 depends on https://github.com/kubernetes/kubernetes/pull/121859 which is still under review. Let's get #121859 merged before attempting this backport.

**Pull request merged**